src/client/gui/gui_cmd.cpp

```diff
@@ -255,6 +278,72 @@ void cmd::GuiCmd::set_menu_actions_for(const std::string& instance_name, const m
     tray_icon_menu.insertMenu(about_separator, instance_menu.get());
 }
 
+void cmd::GuiCmd::handle_petenv_instance(const google::protobuf::RepeatedPtrField<mp::ListVMInstance>& instances)
+{
+    auto petenv_name = Settings::instance().get(petenv_key);
+
+    auto petenv_instance =
+        std::find_if(instances.cbegin(), instances.cend(), [&petenv_name](const ListVMInstance& instance) {
+            return petenv_name.toStdString() == instance.name();
+        });
+
+    if (petenv_instance == instances.cend())
+    {
+        petenv_start_action.setText("Start");
+        petenv_start_action.setEnabled(false);
+        petenv_stop_action.setEnabled(false);
+
+        return;
+    }
+
+    auto state = petenv_instance->instance_status();
+    auto state_string = QString::fromStdString(mp::format::status_string_for(state));
+
+
+    if (petenv_state.status() != state.status())
+    {
+        petenv_start_action.setText(QString::fromStdString(
+            fmt::format("Start \"{}\"{}", petenv_name,
+                        (state.status() != InstanceStatus::STOPPED) ? fmt::format(" ({})", state_string) : "")));
+
+        set_petenv_actions_for(state);
+        petenv_state = state;
+    }
+}
+
+void cmd::GuiCmd::set_petenv_actions_for(const mp::InstanceStatus& state)
+{
+    const auto can_stop_states = {InstanceStatus::UNKNOWN, InstanceStatus::RUNNING, InstanceStatus::DELAYED_SHUTDOWN};
+    const auto can_start_states = {InstanceStatus::STOPPED, InstanceStatus::SUSPENDED};
+
+    if (std::find(can_stop_states.begin(), can_stop_states.end(), state.status()) != can_stop_states.end())
+    {
+        if (state.status() == InstanceStatus::UNKNOWN)
+            petenv_shell_action.setEnabled(false);
+        else
+            petenv_shell_action.setEnabled(true);
+
+        petenv_stop_action.setEnabled(true);
+        petenv_start_action.setEnabled(false);
+    }
+    else if (std::find(can_start_states.begin(), can_start_states.end(), state.status()) != can_start_states.end())
+    {
+        petenv_shell_action.setEnabled(true);
+        petenv_start_action.setEnabled(true);
+        petenv_stop_action.setEnabled(false);
+    }
+    else
+    {
+        if (state.status() == InstanceStatus::DELETED || state.status() == InstanceStatus::SUSPENDING)
```

Comment:
Why SUSPENDING?

Reply:
Why would you "shell" into an instance that is suspending?  The instance is not in a state ready to handle shell, so I don't think "Open Shell" should be available.

Reply:
Indeed, but that is also true of `STOPPING`, right? And `UNKNOWN`, which is handled separately above. 

`stop` and `start` cover complement sets of the possible states (they are mutually exclusive and, together, they cover all of them). This is not true of `shell` (it covers some states from each).So I think the code would be cleaner if the handling of `petenv_shell_action` got separated, with its own `can_shell_states` + if-else. WDYT?

Reply:
... or maybe `cant_shell_states` (would be shorter).

Reply:
Well, there is no `STOPPING` state, although there should be, but I get what you mean here.  I have thought of a different way to handle this, so I'll push that soon.

Reply:
Ok, I updated this section to hopefully be more clear and better capable of handling different states and the action combinations.